chrisbrinkmann/skynet
# Add register new user API route, Setup testing environment, Implement register user tests

### Description

Implement register new user API route,
Setup testing environment,
Implement register new user tests

### Changes

1. Implement handler for register new user api route (/users/register) in `routes/users.js`
  - Added new environment variable (JWT_SECRET) to /config/.env - you will need to add it to your own `.env.local` and `.env.test` as well.

  - Added some packages related to user registration:
    - bcryptjs - password hashing
    - express-validator - server side validation of request data
    - gravatar - stores a profile image at some url
    - jsonwebtoken - token used for user authentication

2. Refactored some functions from the route handler into `utils/utils.js`

3. Setup Jest testing environment for backend:
  - Edit the readme with instructions for setting up and running tests
  - Add npm script for creating test database: `npm createdb-test`
  - Add npm script for running tests: `npm test`
  - Edit `config/jest.config.js` to use `.env.test`
  - Add 'supertest' package for http req testing
  - Refactored `server.js` into `app.js` and `server.js`. This way our tests can access our express app in `app.js` without starting the server our syncing the database outside of the testing environment.
  - Add `db.js` file to /tests/fixtures to setup database environment for tests

4. Implement register new user tests for the /users/register api endpoint

### Verification

You will need to do some setup for these changes to work in your local environment:
- Add `JWT_SECRET` to your `.env.local` file and give it some value. This variable is needed for the Jsonwebtoken creation and verification.
- Follow the steps in the readme under 'Testing the Server' to set up the test database and environment. Specifically you will need to:
  - Create a local `.env.test` file and copy the contents of `.env` into `.env.test`. Set `DB_NAME` to `skynet-test`, and adjust any other variables to your local environment. Make sure this file also has the `JWT_SECRET` variable set.
  - Create a local DB named `skynet-test`... just the script to do this `$ npm run createdb-test`

Now you should be able to use and run test for the /users/register endpoint.

`$npm test`

![image](https://user-images.githubusercontent.com/48368341/66159298-b5b81880-e5f5-11e9-8e32-e988b2e844ba.png)


### Trello Link

https://trello.com/c/cu3HqNl7

### Notes

[other references]
